mlflow/store/sqlalchemy_store.py

```diff
@@ -248,102 +278,104 @@ def _check_run_is_deleted(self, run):
-            # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
-            self._get_or_create(SqlMetric, run_uuid=run_uuid, key=metric.key,
-                                value=metric.value, timestamp=metric.timestamp)
-        except sqlalchemy.exc.IntegrityError as ie:
-            # Querying metrics from run entails pushing the query down to DB layer.
-            # Hence the rollback.
-            self.session.rollback()
-            existing_metric = [m for m in run.metrics
-                               if m.key == metric.key and m.timestamp == metric.timestamp]
-            if len(existing_metric) == 0:
-                raise MlflowException("Log metric request failed for run ID={}. Attempted to log"
-                                      " metric={}. Error={}".format(run_uuid,
-                                                                    (metric.key, metric.value),
-                                                                    str(ie)))
-            else:
-                m = existing_metric[0]
-                raise MlflowException('Metric={} must be unique. Metric already logged value {} '
-                                      'at {}'.format(metric, m.value, m.timestamp),
-                                      INVALID_PARAMETER_VALUE)
+        with self.ManagedSessionMaker() as session:
+            run = self._get_run(run_uuid=run_uuid, session=session)
+            self._check_run_is_active(run)
+            try:
+                # This will check for various integrity checks for metrics table.
+                # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
+                self._get_or_create(model=SqlMetric, run_uuid=run_uuid, key=metric.key,
+                                    value=metric.value, timestamp=metric.timestamp, session=session)
+            except sqlalchemy.exc.IntegrityError as ie:
+                existing_metric = [m for m in run.metrics
+                                   if m.key == metric.key and m.timestamp == metric.timestamp]
+                if len(existing_metric) == 0:
+                    raise MlflowException(
+                        "Log metric request failed for run ID={}. Attempted to log metric={}."
+                        " Error={}".format(run_uuid, (metric.key, metric.value), str(ie)))
+                else:
+                    m = existing_metric[0]
+                    raise MlflowException(
+                        "Metric={} must be unique. Metric already logged value {}"
+                        " at {}".format(metric, m.value, m.timestamp), INVALID_PARAMETER_VALUE)
 
     def get_metric_history(self, run_uuid, metric_key):
-        metrics = self.session.query(SqlMetric).filter_by(run_uuid=run_uuid, key=metric_key).all()
-        return [metric.to_mlflow_entity() for metric in metrics]
+        with self.ManagedSessionMaker() as session:
```

Comment:
Technically, this method does not strictly adhere to the principle that session construction should happen independently of the function that uses the session to communicate with the database (the `ManagedSessionMaker` is used to create a session in the same method that later performs a DB lookup using the session).

We can define an internal `_get_metric_history()` method to achieve this separation, or we can recognize that the separation is a recommended SQLAlchemy pattern for avoiding errors that we may not always follow if we're confident that a certain function is implemented correctly (as I believe is the case here).

What are your thoughts, @mparkhe?

Reply:
IMO, this is fine. Not sure if this principle "session construction should happen independently of the function that uses the session to communicate with the database" makes too much sense from the design perspective. I suppose most of the other functions fit the spec that you might have multiple APIs sharing functions so create the session in the public API for `SqlAlchemyStore` since that is the first entry point.

Ideally, in fact, I might be tempted to have principle of creating a session closets to DB access. However, that might be a bit of an overuse if you have a single call stack needing to make a couple of DB calls to respond to an API request. Also they way you have entrypoint APIs creating sessions is awesome in case we want to share sessions etc. in future. If the same function can make DB calls, so be it.

Reply:
Makes sense to me!